File: crates/breez-sdk/bindings/examples/cli/langs/kotlin-multiplatform/src/main/kotlin/Webhooks.kt
import breez_sdk_spark.*
import org.jline.reader.LineReader

data class WebhooksCliCommand(
    val name: String,
    val description: String,
    val run: suspend (sdk: BreezSdk, reader: LineReader, args: List<String>) -> Unit
)

val WEBHOOKS_COMMAND_NAMES = listOf(
    "register",
    "unregister",
    "list",
)

fun buildWebhooksRegistry(): Map<String, WebhooksCliCommand> {
    return mapOf(
        "register" to WebhooksCliCommand("register", "Register a new webhook", ::handleRegisterWebhook),
        "unregister" to WebhooksCliCommand("unregister", "Unregister a webhook", ::handleUnregisterWebhook),
        "list" to WebhooksCliCommand("list", "List all registered webhooks", ::handleListWebhooks),
    )
}

suspend fun dispatchWebhooksCommand(
    args: List<String>,
    sdk: BreezSdk,
    registry: Map<String, WebhooksCliCommand>,
    reader: LineReader,
) {
    if (args.isEmpty() || args[0] == "help") {
        println()
        println("Webhooks subcommands:")
        registry.keys.sorted().forEach { name ->
            val cmd = registry[name]!!
            println("  webhooks %-26s %s".format(name, cmd.description))
        }
        println()
        return
    }

    val subName = args[0]
    val subArgs = args.drop(1)

    val cmd = registry[subName]
    if (cmd == null) {
        println("Unknown webhooks subcommand: $subName. Use 'webhooks help' for available commands.")
        return
    }

    try {
        cmd.run(sdk, reader, subArgs)
    } catch (e: Exception) {
        println("Error: ${e.message}")
    }
}

// ---------------------------------------------------------------------------
// Webhooks command handlers
// ---------------------------------------------------------------------------

suspend fun handleRegisterWebhook(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    val fp = FlagParser(args)
    val url = fp.getString("url") ?: fp.positional.getOrNull(0)
    val secret = fp.getString("secret") ?: fp.positional.getOrNull(1)
    val eventsStr = fp.getString("events") ?: fp.positional.drop(2).joinToString(",").ifEmpty { null }

    if (url == null || secret == null || eventsStr == null) {
        println("Usage: webhooks register --url <url> --secret <secret> --events <type1,type2,...>")
        println("Event types: lightning-receive, lightning-send, coop-exit, static-deposit")
        return
    }

    val eventTypes = eventsStr.split(",").map { s ->
        when (s.trim().lowercase()) {
            "lightning-receive" -> WebhookEventType.LightningReceiveFinished
            "lightning-send" -> WebhookEventType.LightningSendFinished
            "coop-exit" -> WebhookEventType.CoopExitFinished
            "static-deposit" -> WebhookEventType.StaticDepositFinished
            else -> {
                println("Unknown event type: ${s.trim()}")
                return
            }
        }
    }

    val result = sdk.registerWebhook(
        RegisterWebhookRequest(
            url = url,
            secret = secret,
            eventTypes = eventTypes,
        )
    )
    printValue(result)
}

suspend fun handleUnregisterWebhook(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    if (args.isEmpty()) {
        println("Usage: webhooks unregister <webhook_id>")
        return
    }

    sdk.unregisterWebhook(UnregisterWebhookRequest(webhookId = args[0]))
    println("Webhook unregistered successfully")
}

suspend fun handleListWebhooks(sdk: BreezSdk, reader: LineReader, args: List<String>) {
    val result = sdk.listWebhooks()
    printValue(result)
}

File: crates/breez-sdk/bindings/examples/cli/langs/swift/Sources/BreezCLI/Webhooks.swift
import Foundation
import BreezSdkSpark

// MARK: - Webhooks command names (for REPL completion)

let webhooksCommandNames: [String] = [
    "webhooks register",
    "webhooks unregister",
    "webhooks list",
]

// MARK: - Dispatch

func dispatchWebhooksCommand(_ args: [String], sdk: BreezSdk) async {
    if args.isEmpty || args[0] == "help" {
        printWebhooksHelp()
        return
    }

    let subName = args[0]
    let subArgs = Array(args.dropFirst())

    do {
        switch subName {
        case "register":
            try await handleWebhookRegister(sdk, subArgs)
        case "unregister":
            try await handleWebhookUnregister(sdk, subArgs)
        case "list":
            try await handleWebhookList(sdk, subArgs)
        default:
            print("Unknown webhooks subcommand: \(subName). Use 'webhooks help' for available commands.")
        }
    } catch {
        print("Error: \(error)")
    }
}

// MARK: - Help

private func printWebhooksHelp() {
    print("\nWebhooks subcommands:")
    print("  webhooks \("register".padding(toLength: 30, withPad: " ", startingAt: 0))Register a new webhook")
    print("  webhooks \("unregister".padding(toLength: 30, withPad: " ", startingAt: 0))Unregister a webhook")
    print("  webhooks \("list".padding(toLength: 30, withPad: " ", startingAt: 0))List all registered webhooks")
    print()
}

// MARK: - Parsing helper

private func parseWebhookEventType(_ raw: String) -> WebhookEventType? {
    switch raw.lowercased().replacingOccurrences(of: "-", with: "").replacingOccurrences(of: "_", with: "") {
    case "lightningreceive": return .lightningReceiveFinished
    case "lightningsend": return .lightningSendFinished
    case "coopexit": return .coopExitFinished
    case "staticdeposit": return .staticDepositFinished
    default: return nil
    }
}

// MARK: - Handlers

// --- register ---

private func handleWebhookRegister(_ sdk: BreezSdk, _ args: [String]) async throws {
    let fp = FlagParser(args)
    guard fp.positional.count >= 3 else {
        print("Usage: webhooks register <url> <secret> <event_type> [<event_type> ...]")
        print("Event types: lightning-receive, lightning-send, coop-exit, static-deposit")
        return
    }

    let url = fp.positional[0]
    let secret = fp.positional[1]
    let eventTypes = fp.positional.dropFirst(2).compactMap { parseWebhookEventType(String($0)) }

    if eventTypes.isEmpty {
        print("No valid event types provided")
        print("Event types: lightning-receive, lightning-send, coop-exit, static-deposit")
        return
    }

    let result = try await sdk.registerWebhook(request: RegisterWebhookRequest(
        url: url,
        secret: secret,
        eventTypes: eventTypes
    ))
    printValue(result)
}

// --- unregister ---

private func handleWebhookUnregister(_ sdk: BreezSdk, _ args: [String]) async throws {
    guard !args.isEmpty else {
        print("Usage: webhooks unregister <webhook_id>")
        return
    }

    try await sdk.unregisterWebhook(request: UnregisterWebhookRequest(
        webhookId: args[0]
    ))
    print("Webhook unregistered successfully")
}

// --- list ---

private func handleWebhookList(_ sdk: BreezSdk, _ args: [String]) async throws {
    let result = try await sdk.listWebhooks()
    printValue(result)
}
